### PR TITLE
Fix replace in visual, visual line, and visual block mode

### PR DESCRIPTION
**Pull request by xconverge**
Any feedback on better ways of doing any of this?

1. Visual block mode returns correctly after running "r" instead of having a few cursors left around
2. Replace does not replace unless there is a character, this applies to line, block, and regular visual
3. Replace added to visual and visual line with ~~an IterateSelection position function~~ selective replace transforms in chunks

### CODE REVIEW

File: src/mode/modeHandler.ts [hunk 1]
```diff
@@ -1366,6 +1366,9 @@ export class ModeHandler implements vscode.Disposable {
             Position.EarlierOf(start, stop).getLineBegin(),
             Position.LaterOf(start, stop).getLineEnd()
           ) ];
+          vimState.cursorStartPosition = selections[0].start as Position;
```
Comment (xconverge) [inline src/mode/modeHandler.ts]: This is updating cursor positions correctly for visual line mode

File: src/motion/position.ts [hunk 2]
```diff
@@ -192,6 +192,41 @@ export class Position extends vscode.Position {
   }
 
   /**
+   * Iterate over every position in the selection defined by the two positions passed in.
+   */
+  public static *IterateSelection(topLeft: Position, bottomRight: Position): Iterable<{ line: string, char: string, pos: Position }> {
```
Comment (xconverge) [inline src/motion/position.ts]: I am not using it anymore but I will leave it since it seems handy

File: src/actions/actions.ts [hunk 3]
```diff
@@ -3838,6 +3838,86 @@ class ActionReplaceCharacter extends BaseCommand {
...
 
 @RegisterAction
+class ActionReplaceCharacterVisual extends BaseCommand {
+  modes = [ModeName.Visual, ModeName.VisualLine];
+  keys = ["r", "<character>"];
+  runsOnceForEveryCursor() { return false; }
+  canBeRepeatedWithDot = true;
+
+  public async exec(position: Position, vimState: VimState): Promise<VimState> {
+    const toInsert   = this.keysPressed[1];
```
Comment (rebornix) [inline src/actions/actions.ts]: redundant spaces.

File: src/actions/actions.ts [hunk 4]
```diff
@@ -3838,6 +3838,86 @@ class ActionReplaceCharacter extends BaseCommand {
...
+
+    let visualSelectionOffset = 1;
+    let start = vimState.cursorStartPosition;
+    let end = vimState.cursorPosition;
+
+    // If selection is reversed, reorganize it so that the text replace logic always works
+    if (start.line > end.line || ((start.line === end.line)
+      && (start.character > end.character))) {
+      let tmp = start;
```
Comment (xconverge) [inline src/actions/actions.ts]: Ah dang, @johnfn guided me to this within the past month... will do

File: src/actions/actions.ts (hunk 3 above)
Comment (xconverge) [inline src/actions/actions.ts]: fixed

File: src/actions/actions.ts [hunk 5]
```diff
@@ -3838,6 +3838,86 @@ class ActionReplaceCharacter extends BaseCommand {
...
+
+  public async exec(position: Position, vimState: VimState): Promise<VimState> {
+    const toInsert   = this.keysPressed[1];
+
+    let visualSelectionOffset = 1;
+    let start = vimState.cursorStartPosition;
+    let end = vimState.cursorPosition;
+
+    // If selection is reversed, reorganize it so that the text replace logic always works
+    if (start.line > end.line || ((start.line === end.line)
```
Comment (xconverge) [inline src/actions/actions.ts]: fixed

File: src/actions/actions.ts (hunk 4 above)
Comment (xconverge) [inline src/actions/actions.ts]: fixed

**Comment by xconverge** [comment]
Should be good to go now

thanks for the EOL find @Platzer

**Pull request merged**